# Machine Learning Classification

*Author: Evan Carey*

*Copyright 2017-2019, BH Analytics, LLC*

------------------------------------------------
*Adjusted for HDS-5230 Week9 Assignmnent*

*Adjusted by Wenshan, Liu*

*Date: 04-01-2025*

## Overview

The purpose of this section is to go over machine learning! We will focus on classification in the context of python (the scikit-learn module). We will include some general concepts of machine learning as well as the specifics of a few different classification algorithms. For further reading, I highly recommend the free ebook titled 'Introduction to Statistical Learning' by Gareth James. A quick web search should find this book near the top of the search results. For even more in-depth coverage of machine learning algorithms, I recommend the book  'Elements of Statistical Learning' by Trevor Hastie (also free online). 

## Classification

In the case where our outcome (target) variable is discrete with a limited number of possible values, we can use classification algorithms to predict the outcome. Imagine a binary outcome with values of 'Yes' and 'No'. We are interested in predicting the probability that the outcome is either 'Yes' or 'No'. It is also possible to predict outcomes with more than two possible values, but we will focus on the binary case here. 

## Libraries

In [78]:
## Import Modules
import os
import sys
import numpy as np
import pandas as pd
import matplotlib
import matplotlib.pyplot as plt
from patsy import dmatrices
from sklearn.metrics import confusion_matrix
import sklearn
from sklearn import datasets

In [79]:
#!pip install patsy

In [80]:
## Set default figure size to be larger 
## this may only work in matplotlib 2.0+!
matplotlib.rcParams['figure.figsize'] = [10.0,6.0]
## Enable multiple outputs from jupyter cells
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

In [81]:
## Get Version information
print(sys.version)
print("Pandas version: {0}".format(pd.__version__))
print("Matplotlib version: {0}".format(matplotlib.__version__))
print("Numpy version: {0}".format(np.__version__))
print("SciKitLearn version: {0}".format(sklearn.__version__))

3.10.14 (main, May  6 2024, 14:42:37) [Clang 14.0.6 ]
Pandas version: 2.2.2
Matplotlib version: 3.9.2
Numpy version: 2.1.0
SciKitLearn version: 1.5.1


## Check your working directory

Set your working directory to make paths easier :) 

In [82]:
# Working Directory
import os
print("My working directory:\n" + os.getcwd())
# Set Working Directory 
os.chdir(".")
print("My new working directory:\n" + os.getcwd())

My working directory:
/Users/sandyliu/Library/CloudStorage/OneDrive-Personal/2025_Spring/HDS5230-Assignment/W9-Assignment
My new working directory:
/Users/sandyliu/Library/CloudStorage/OneDrive-Personal/2025_Spring/HDS5230-Assignment/W9-Assignment


## Patient Mortality Dataset

We will use a dataset with a binary outcome of mortality as a motivating example.

This is a dataset of patients demographics and disease status, with mortality indicated. The dataset is here: 

`data\healthcare\patientAnalyticFile.csv`

In practice, you most likely would have created a dataset like this from multiple other files after cleaning, reshaping, and joining them. 

You can generalize this setup to any situation with a binary outcome, such as estimating the probability of a customer filing a warranty claim, or the probability of a transaction being fraudulent. 

We will first import this dataset and examine the potential variables to use in our classification algorithm.

In [83]:
## Set print limits
pd.options.display.max_rows = 10
## Import Data
df_patient = \
 pd.read_csv('./PatientAnalyticFile.csv')
df_patient

PatientID DateOfBirth  Gender      Race  Myocardial_infarction  \
0              1  1962-02-27  female  hispanic                      0   
1              2  1959-08-18    male     white                      0   
2              3  1946-02-15  female     white                      0   
3              4  1979-07-27  female     white                      0   
4              5  1983-02-19  female  hispanic                      0   
...          ...         ...     ...       ...                    ...   
19995      19996  1997-12-19  female     other                      0   
19996      19997  1984-03-31  female     white                      0   
19997      19998  1993-07-04  female     white                      0   
19998      19999  1984-04-17    male     other                      0   
19999      20000  1966-05-14  female     white                      0   

       Congestive_heart_failure  Peripheral_vascular_disease  Stroke  \
0                             0                            0       0   
1                             0                            0       0   
2                             0                            0       0   
3                             0                            0       0   
4                             0                            0       0   
...                         ...                          ...     ...   
19995                         0                            0       0   
19996                         0                            0       0   
19997                         0                            0       0   
19998                         0                            0       0   
19999                         0                            0       0   

       Dementia  Pulmonary  ...  Metastatic_solid_tumour  HIV  Obesity  \
0             0          0  ...                        0    0        0   
1             0          0  ...                        0    0        0   
2             0          0  ...                        0    1        0   
3             0          1  ...                        0    0        0   
4             0          0  ...                        0    0        0   
...         ...        ...  ...                      ...  ...      ...   
19995         0          0  ...                        0    0        0   
19996         0          0  ...                        0    1        0   
19997         0          0  ...                        0    0        1   
19998         0          0  ...                        0    0        0   
19999         0          0  ...                        0    0        0   

       Depression  Hypertension  Drugs  Alcohol  First_Appointment_Date  \
0               0             0      0        0              2013-04-27   
1               0             1      0        0              2005-11-30   
2               0             1      0        0              2011-11-05   
3               0             0      0        0              2010-03-01   
4               0             1      0        0              2006-09-22   
...           ...           ...    ...      ...                     ...   
19995           0             0      0        0              2008-06-14   
19996           0             1      0        0              2007-04-24   
19997           0             1      0        0              2010-10-16   
19998           0             1      0        0              2015-01-04   
19999           0             0      1        0              2011-04-01   

       Last_Appointment_Date  DateOfDeath  
0                 2018-06-01          NaN  
1                 2008-11-02   2008-11-02  
2                 2015-11-13          NaN  
3                 2016-01-17   2016-01-17  
4                 2018-06-01          NaN  
...                      ...          ...  
19995             2018-06-01          NaN  
19996             2018-06-01          NaN  
19997             2018-06-01          NaN  
19998             2018-06-01       

We need to make a variable to indicate mortality. We can do that based on the abscence of 'date of death':

In [84]:
# Create mortality variable
df_patient['mortality'] = \
    np.where(df_patient['DateOfDeath'].isnull(),
             0,1)
# Examine
df_patient['mortality']

0        0
1        1
2        0
3        1
4        0
        ..
19995    0
19996    0
19997    0
19998    0
19999    0
Name: mortality, Length: 20000, dtype: int64

In [85]:
df_patient['mortality'].describe()

count    20000.000000
mean         0.354700
std          0.478434
min          0.000000
25%          0.000000
50%          0.000000
75%          1.000000
max          1.000000
Name: mortality, dtype: float64

In [86]:
df_patient.describe()

PatientID  Myocardial_infarction  Congestive_heart_failure  \
count  20000.000000           20000.000000              20000.000000   
mean   10000.500000               0.045600                  0.043450   
std     5773.647028               0.208621                  0.203873   
min        1.000000               0.000000                  0.000000   
25%     5000.750000               0.000000                  0.000000   
50%    10000.500000               0.000000                  0.000000   
75%    15000.250000               0.000000                  0.000000   
max    20000.000000               1.000000                  1.000000   

       Peripheral_vascular_disease        Stroke      Dementia     Pulmonary  \
count                 20000.000000  20000.000000  20000.000000  20000.000000   
mean                      0.023950      0.028650      0.031400      0.072650   
std                       0.152897      0.166825      0.174401      0.259568   
min                       0.000000      0.000000      0.000000      0.000000   
25%                       0.000000      0.000000      0.000000      0.000000   
50%                       0.000000      0.000000      0.000000      0.000000   
75%                       0.000000      0.000000      0.000000      0.000000   
max                       1.000000      1.000000      1.000000      1.000000   

          Rheumatic  Peptic_ulcer_disease     LiverMild  ...        Cancer  \
count  20000.000000          20000.000000  20000.000000  ...  20000.000000   
mean       0.012300              0.009650      0.009250  ...      0.050450   
std        0.110224              0.097762      0.095733  ...      0.218877   
min        0.000000              0.000000      0.000000  ...      0.000000   
25%        0.000000              0.000000      0.000000  ...      0.000000   
50%        0.000000              0.000000      0.000000  ...      0.000000   
75%        0.000000              0.000000      0.000000  ...      0.000000   
max        1.000000              1.000000      1.000000  ...      1.000000   

        LiverSevere  Metastatic_solid_tumour           HIV       Obesity  \
count  20000.000000             20000.000000  20000.000000  20000.000000   
mean       0.051450                 0.033150      0.006450      0.163450   
std        0.220919                 0.179033      0.080054      0.369785   
min        0.000000                 0.000000      0.000000      0.000000   
25%        0.000000                 0.000000      0.000000      0.000000   
50%        0.000000                 0.000000      0.000000      0.000000   
75%        0.000000                 0.000000      0.000000      0.000000   
max        1.000000                 1.000000      1.000000      1.000000   

         Depression  Hypertension         Drugs       Alcohol     mortality  
count  20000.000000  20000.000000  20000.000000  20000.000000  20000.000000  
mean       0.106300      0.302900      0.040050      0.079750      0.354700  
std        0.308229      0.459524      0.196081      0.270913      0.478434  
min        0.000000      0.000000      0.000000      0.000000      0.000000  
25%        0.000000      0.000000      0.000000      0.000000      0.000000  
50%        0.000000      0.000000      0.000000      0.000000      0.000000  
75%        0.000000      1.000000      0.000000      0.000000      1.000000  
max        1.000000      1.000000      1.000000      1.000000      1.000000  

[8 rows x 24 columns]

In [87]:
df_patient.dtypes

PatientID                  int64
DateOfBirth               object
Gender                    object
Race                      object
Myocardial_infarction      int64
                           ...  
Alcohol                    int64
First_Appointment_Date    object
Last_Appointment_Date     object
DateOfDeath               object
mortality                  int64
Length: 30, dtype: object

We should change date of birth to be an actual date and calculate age if we want to include it in the model:

In [88]:
# Convert dateofBirth to date
df_patient['DateOfBirth'] = \
    pd.to_datetime(df_patient['DateOfBirth'])
# Calculate age in years as of 2015-01-01
df_patient['Age_years'] = \
    ((pd.to_datetime('2015-01-01') - df_patient['DateOfBirth']).dt.days/365.25)
df_patient['Age_years'].describe()

count    20000.000000
mean        47.247474
std         18.145086
min         15.753593
25%         31.733744
50%         47.099247
75%         62.924025
max         78.743326
Name: Age_years, dtype: float64

## Workflow into scikit-learn


* There are a number of possible ways to prepare data for modeling in scikit-learn. 
* You must end up with a numeric ndarray of inputs (X) and a numeric ndarray matrix of the target (Y)
* I prefer the following workflow:
  * We use pandas to import and clean data
  * We use Patsy to create the X and Y ndarrays
  * Using categorical transformations (dummy coding) as needed
  * Also can generate non-linear terms including splines
  * Use scikit-learn for machine learning

## Use Patsy to Create the Model Matrices

We typically start out with a pandas dataframe for manipulation purposes, then we will use this dataframe as the input to the machine learning library. I created a pandas dataframe above to replicate this process. We will use the dmatrices function from the patsy library to easily generate the design matrices for the machine learning algorithms representing the inputs. THis handles the following:

* drops rows with missing data
* construct one-hot encoding for categorical variables
* optionally adds constant intecercept

In [89]:
df_patient.columns

Index(['PatientID', 'DateOfBirth', 'Gender', 'Race', 'Myocardial_infarction',
       'Congestive_heart_failure', 'Peripheral_vascular_disease', 'Stroke',
       'Dementia', 'Pulmonary', 'Rheumatic', 'Peptic_ulcer_disease',
       'LiverMild', 'Diabetes_without_complications',
       'Diabetes_with_complications', 'Paralysis', 'Renal', 'Cancer',
       'LiverSevere', 'Metastatic_solid_tumour', 'HIV', 'Obesity',
       'Depression', 'Hypertension', 'Drugs', 'Alcohol',
       'First_Appointment_Date', 'Last_Appointment_Date', 'DateOfDeath',
       'mortality', 'Age_years'],
      dtype='object')

In [90]:
## Create formula for all variables in model
vars_remove = ['PatientID','First_Appointment_Date','DateOfBirth',
               'Last_Appointment_Date','DateOfDeath','mortality']
vars_left = set(df_patient.columns) - set(vars_remove)
formula = "mortality ~ " + " + ".join(vars_left)
formula

'mortality ~ Gender + HIV + LiverMild + Congestive_heart_failure + Myocardial_infarction + Depression + Diabetes_without_complications + Peptic_ulcer_disease + Peripheral_vascular_disease + Renal + Pulmonary + Cancer + Diabetes_with_complications + Alcohol + Drugs + Rheumatic + LiverSevere + Stroke + Obesity + Dementia + Paralysis + Metastatic_solid_tumour + Age_years + Hypertension + Race'

In [91]:
## only use subset of data so models fit in reasonable time
df_patient_sub = \
    df_patient.sample(frac=0.1,
                     random_state=32)    
## use Patsy to create model matrices
Y,X = dmatrices(formula,
                df_patient_sub)

In [92]:
Y

DesignMatrix with shape (2000, 1)
  mortality
          0
          0
          1
          1
          0
          0
          1
          1
          0
          0
          1
          0
          1
          0
          1
          0
          1
          0
          0
          1
          0
          1
          0
          0
          0
          0
          1
          1
          0
          0
  [1970 rows omitted]
  Terms:
    'mortality' (column 0)
  (to view full data, use np.asarray(this_obj))

In [93]:
X

DesignMatrix with shape (2000, 28)
  Columns:
    ['Intercept',
     'Gender[T.male]',
     'Race[T.hispanic]',
     'Race[T.other]',
     'Race[T.white]',
     'HIV',
     'LiverMild',
     'Congestive_heart_failure',
     'Myocardial_infarction',
     'Depression',
     'Diabetes_without_complications',
     'Peptic_ulcer_disease',
     'Peripheral_vascular_disease',
     'Renal',
     'Pulmonary',
     'Cancer',
     'Diabetes_with_complications',
     'Alcohol',
     'Drugs',
     'Rheumatic',
     'LiverSevere',
     'Stroke',
     'Obesity',
     'Dementia',
     'Paralysis',
     'Metastatic_solid_tumour',
     'Age_years',
     'Hypertension']
  Terms:
    'Intercept' (column 0)
    'Gender' (column 1)
    'Race' (columns 2:5)
    'HIV' (column 5)
    'LiverMild' (column 6)
    'Congestive_heart_failure' (column 7)
    'Myocardial_infarction' (column 8)
    'Depression' (column 9)
    'Diabetes_without_complications' (column 10)
    'Peptic_ulcer_disease' (column 11)
    'Perip

## Split into Testing and Training Samples

* The first step is to set aside a test sample of data that will allow us to estimate the generalization error post-fit. This protects against overfitting. 
* We can use “tuple unpacking” to assign the values (very pythonic :)
* We can assign a random seed (state) and fraction to split.

 For simple random splits, scikit-learn has a function `train_test_split()`

In [94]:
## Split Data into training and sample
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = \
    train_test_split(X,
                     np.ravel(Y), # prevents dimensionality error later!
                     test_size=0.25,
                     random_state=42)

## Confirm the Output Dimensions

* We can confirm the dimensions of the data are the same within test and train
* The proportion should also be close to the test_size argument. 

In [95]:
## Confirm dimensions
X_train.shape

(1500, 28)

In [96]:
X_test.shape

(500, 28)

In [97]:
y_train.shape

(1500,)

In [98]:
y_test.shape

(500,)

## First Model: Logistic Regression

* We will start with a basic logistic regression model.
* The flow will be similar for other models
* Call and save model object with initial parameters, then call the fit() method to perform the optimization
* Then call other summary methods post fit to explore the model

Check the docs: 

https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html

In [99]:
## import linear model
from sklearn import linear_model
## Define model parameters
## can implement penalties, but check docs for appropriate solver
clf = linear_model.LogisticRegression(fit_intercept=True, # already have the intercept
                                      solver='liblinear') # could change to lbfgs!
## fit model using data with .fit
clf.fit(X_train,y_train)

LogisticRegression(solver='liblinear')

How do we know if this is a good model? What makes a good model? Let's make predictions, is this a good model? Which parameters are most important?

In [100]:
## Make predictions on training dataset
## training error?
clf.predict(X_train)

array([0., 0., 0., ..., 0., 0., 0.])

In [101]:
## can also predict probabilities
clf.predict_proba(X_train)

array([[0.7146111 , 0.2853889 ],
       [0.72980989, 0.27019011],
       [0.91840587, 0.08159413],
       ...,
       [0.91934318, 0.08065682],
       [0.93820382, 0.06179618],
       [0.72821741, 0.27178259]])

## Model Summaries

* We can extract the model coefficients with the .coef_ attribute.

In [102]:
## Get coefficients
clf.coef_
clf.coef_.shape

array([[-1.8624683 , -0.10167484, -0.11595604, -0.18726502, -0.0453132 ,
         0.09894352,  0.16254145,  0.40421821,  0.5144615 ,  0.56220815,
        -0.26651427, -0.34338786,  0.87534653,  0.27342733,  0.33183451,
         0.42038745, -0.63768785,  1.03493315,  0.3111227 ,  0.40896697,
         0.71968793,  0.1008458 ,  0.25592983, -0.01869836, -0.01891693,
        -0.2937289 ,  0.05813761, -0.02121856]])

(1, 28)

## Assessing the Model Score (accuracy)

* The concept of error is a bit different for a binary outcome than the continuous case. 
* We can construct error to be a function of the number of incorrect predictions. 

In [103]:
## get mean accuracy
clf.score(X_train,y_train) 

0.7333333333333333

In this case, our model is accurate about 73% of the time! We can understand what that means by looking at the predictions against the actual outcomes. This is called a confusion matrix.

| True negative  | False positive |
|----------------|----------------|
| False negative | True positive |

In [104]:
## get confusion matrix
from sklearn.metrics import confusion_matrix
confusion_matrix(y_train,
                 clf.predict(X_train))

array([[826, 144],
       [256, 274]])

In [105]:
## get classification metrics
print(sklearn.metrics.classification_report(y_train,
                                            clf.predict(X_train)))

              precision    recall  f1-score   support

         0.0       0.76      0.85      0.81       970
         1.0       0.66      0.52      0.58       530

    accuracy                           0.73      1500
   macro avg       0.71      0.68      0.69      1500
weighted avg       0.73      0.73      0.72      1500



In [106]:
## get accuracy
sklearn.metrics.accuracy_score(y_train,
                               clf.predict(X_train))

0.7333333333333333

Another option for assessing our model is to use the Kappa statistic (instead of accuracy). The Kappa statistics is a measure of rater agreement, with values between -1 and 1. 

* A value of 0 indicates the classifier is not better than chance
* A value of 1 indicates the classifier is a perfect predictor
* A value of -1 indicates the classifier is always wrong!

In [107]:
# Get kappa
sklearn.metrics.cohen_kappa_score(y_train,
                                  clf.predict(X_train))

np.float64(0.3870796387856005)

## Keep Track of Scores Across Models

I am going to write a small function that will print the scores from a dict so we can compare the models. I will store the model scores in the dict as well. 

In [108]:
## Create dict to store all these results:
result_scores = {}
## Score the Model on Training and Testing Set
result_scores['Logistic'] = \
            (sklearn.metrics.accuracy_score(y_train,clf.predict(X_train)),
             sklearn.metrics.accuracy_score(y_test,clf.predict(X_test)))

In [109]:
## Create Function to Print Results
def get_results(x1):
    print("\n{0:20}   {1:4}    {2:4}".format('Model','Train','Test'))
    print('-------------------------------------------')
    for i in x1.keys():
        print("{0:20}   {1:<6.4}   {2:<6.4}".format(i,x1[i][0],x1[i][1]))

In [110]:
get_results(result_scores)


Model                  Train    Test
-------------------------------------------
Logistic               0.7333   0.718 


## Comparison to the Null Model

Is that a good score for accuracy? Compared to what? We can consider a null model of simply predicting the most frequent class as a base model. Without any other information, I may predict based simply on the distribution of the outcome.

In [111]:
## Null information rate
1 - y_train.mean()

array(0.64666667)

Scikitlearn has a built in dummy classifier that works similarly:

In [112]:
## Dummy classifier
from sklearn.dummy import DummyClassifier
clf = DummyClassifier(strategy='most_frequent',
                      random_state=0)
clf.fit(X_train, y_train)
clf.score(X_train, y_train)  

DummyClassifier(random_state=0, strategy='most_frequent')

0.6466666666666666

In [113]:
## Score the Model on Training and Testing Set
result_scores['Null'] = \
            (sklearn.metrics.accuracy_score(y_train,clf.predict(X_train)),
             sklearn.metrics.accuracy_score(y_test,clf.predict(X_test)))

In [114]:
get_results(result_scores)


Model                  Train    Test
-------------------------------------------
Logistic               0.7333   0.718 
Null                   0.6467   0.608 


## Regularized Linear Regression

- The next family of models we will consider are called regularized linear regression. 
- This includes LASSO, Elastic Net, and Ridge regression. 
- These are penalized forms of a regular linear regression (like a logistic regression). 
- The basic idea is that we can place a penalty on the estimated coefficients from the general linear model, 'pushing' them towards zero. 
- If the coefficients are related to the outcome, they will 'push' back against our penalty. 
- The stronger the relationship (or the stronger the predictor), the stronger they will 'push' back. 
- The overall effect is that the coefficients are all shrunk towards zero. If the variable is not strongly related to the outcome, it will be shrunk close to zero, or possibly all the way to zero. 
- This can give us effective variable selection, where the weak variables are eliminated since their coefficients are shrunk all the way to zero. 
- Depending on how we apply the penalty, variables will either be shrunk all the way to zero (this is called the LASSO), or they will be shrunk to a small number, but still above zero (This is called ridge regression).
- We can also apply a mixture of the two penalties, which is called the elastic net regression. 
- A natural question you might ask is, how do I pick the best model?
    + LASSO?
    + Ridge regression?
    + Elastic net (the mixture of the two)?
- Also, how strong of a penalty should I pick?
    + A very weak penalty, so it is essentially just a logistic regression?
    + A very strong penalty, so almost all the coefficients are equal to zero?
    + Maybe something in between?

## Logistic regression with L1 penalty

If we implement an L1 penalty using the logistic regression function, we are implementing a LASSO regression. Under an L2 penalty, coefficients can actually be set all the way to 0, thus they are eliminated (spare models, feature selection). It is called L1 because the penalty is linked to the **absolute value** of the coefficient. From the scikit-learn docs, here is the cost function:

$$\min_{w, c} \|w\|_1 + C \sum_{i=1}^n \log(\exp(- y_i (X_i^T w + c)) + 1).$$

In [115]:
## Logistic Regression with l1 penalty
## Specify penalty directly as C = 1
clf = linear_model.LogisticRegression(penalty='l1',
                                      C=1, solver = 'liblinear') # specify penalty
clf.fit(X_train,y_train)
## get confusion matrix
confusion_matrix(y_train,clf.predict(X_train))

LogisticRegression(C=1, penalty='l1', solver='liblinear')

array([[820, 150],
       [252, 278]])

In [116]:
## get accuracy
sklearn.metrics.accuracy_score(y_train,clf.predict(X_train))

0.732

In [117]:
## Get kappa
sklearn.metrics.cohen_kappa_score(y_train,clf.predict(X_train))

np.float64(0.3867713460521498)

In [118]:
## get classification metrics
print(sklearn.metrics.classification_report(y_train,
                                            clf.predict(X_train)))

              precision    recall  f1-score   support

         0.0       0.76      0.85      0.80       970
         1.0       0.65      0.52      0.58       530

    accuracy                           0.73      1500
   macro avg       0.71      0.68      0.69      1500
weighted avg       0.72      0.73      0.72      1500



In [119]:
## Score the Model on Training and Testing Set
result_scores['Logistic_L1_C_1'] = \
            (sklearn.metrics.accuracy_score(y_train,clf.predict(X_train)),
             sklearn.metrics.accuracy_score(y_test,clf.predict(X_test)))
get_results(result_scores)


Model                  Train    Test
-------------------------------------------
Logistic               0.7333   0.718 
Null                   0.6467   0.608 
Logistic_L1_C_1        0.732    0.716 


What about other values for C? We could try 0.1, or 10?

In [120]:
## Logistic Regression with l1 penalty
## Specify penalty directly as C = 0.1
clf = linear_model.LogisticRegression(penalty='l1',
                                      C=0.1, solver = 'liblinear') # specify penalty
clf.fit(X_train,y_train)
## Score the Model on Training and Testing Set
result_scores['Logistic_L1_C_01'] = \
            (sklearn.metrics.accuracy_score(y_train,clf.predict(X_train)),
             sklearn.metrics.accuracy_score(y_test,clf.predict(X_test)))
get_results(result_scores)

LogisticRegression(C=0.1, penalty='l1', solver='liblinear')


Model                  Train    Test
-------------------------------------------
Logistic               0.7333   0.718 
Null                   0.6467   0.608 
Logistic_L1_C_1        0.732    0.716 
Logistic_L1_C_01       0.726    0.706 


In [121]:
## Logistic Regression with l1 penalty
## Specify penalty directly as C = 0.1
clf = linear_model.LogisticRegression(penalty='l1',
                                      C=10, solver = 'liblinear') # specify penalty
clf.fit(X_train,y_train)
## Score the Model on Training and Testing Set
result_scores['Logistic_L1_C_10'] = \
            (sklearn.metrics.accuracy_score(y_train,clf.predict(X_train)),
             sklearn.metrics.accuracy_score(y_test,clf.predict(X_test)))
get_results(result_scores)

LogisticRegression(C=10, penalty='l1', solver='liblinear')


Model                  Train    Test
-------------------------------------------
Logistic               0.7333   0.718 
Null                   0.6467   0.608 
Logistic_L1_C_1        0.732    0.716 
Logistic_L1_C_01       0.726    0.706 
Logistic_L1_C_10       0.7347   0.718 


## Selecting Parameters via Cross Validation

We should be validating this parameter 'C' somehow. We should not be using the test data to do that however! We need another dataset, called the validation dataset. We could further split our training data into validation and training. Another option would be to implement k-folds cross validation. 

In [122]:
## Select the alpha through cross validation (k-folds leave one out)
# auto generate 20 values between 1e-4 and 1e4 on log scale
clf = linear_model.LogisticRegressionCV(cv=5,
                                        Cs=20, ## takes awhile to fit 20 models!
                                        penalty='l1',
                                        solver='liblinear') 
clf.fit(X_train,y_train)

LogisticRegressionCV(Cs=20, cv=5, penalty='l1', solver='liblinear')

In [123]:
## how many C's were fit?
clf.Cs
## which C's were fit?
clf.Cs_
## Which C was 'best'? 
clf.C_

20

array([1.00000000e-04, 2.63665090e-04, 6.95192796e-04, 1.83298071e-03,
       4.83293024e-03, 1.27427499e-02, 3.35981829e-02, 8.85866790e-02,
       2.33572147e-01, 6.15848211e-01, 1.62377674e+00, 4.28133240e+00,
       1.12883789e+01, 2.97635144e+01, 7.84759970e+01, 2.06913808e+02,
       5.45559478e+02, 1.43844989e+03, 3.79269019e+03, 1.00000000e+04])

array([0.08858668])

In [124]:
## Score the Model on Training and Testing Set
result_scores['Logistic_L1_C_auto'] = \
            (sklearn.metrics.accuracy_score(y_train,clf.predict(X_train)),
             sklearn.metrics.accuracy_score(y_test,clf.predict(X_test)))
get_results(result_scores)


Model                  Train    Test
-------------------------------------------
Logistic               0.7333   0.718 
Null                   0.6467   0.608 
Logistic_L1_C_1        0.732    0.716 
Logistic_L1_C_01       0.726    0.706 
Logistic_L1_C_10       0.7347   0.718 
Logistic_L1_C_auto     0.7233   0.708 


## Scaling / Pipeline

* We should consider scaling when we use regularization methods. 
* We can construct a pipeline to avoid having to apply the same transformation over and over again.
* We must use the StandardScaler() function here.

In [125]:
## LASSO (L1) regression, set alpha
from sklearn import preprocessing
from sklearn.pipeline import Pipeline
## set our transformation
scaler = preprocessing.StandardScaler()

## set our model
clf = linear_model.LogisticRegressionCV(cv=5,
                                        Cs=20, ## takes awhile to fit 20 models!
                                        penalty='l1',
                                        solver='liblinear') 
## put together in pipeline
pipe1 = Pipeline([("scale", scaler),
                  ("LASSO", clf)])
pipe1.fit(X_train,y_train)

Pipeline(steps=[('scale', StandardScaler()),
                ('LASSO',
                 LogisticRegressionCV(Cs=20, cv=5, penalty='l1',
                                      solver='liblinear'))])

You can extract the elements from the pipline using their names. By calling `named_steps`, you get a dict of the steps:

In [126]:
pipe1.named_steps

{'scale': StandardScaler(),
 'LASSO': LogisticRegressionCV(Cs=20, cv=5, penalty='l1', solver='liblinear')}

In [127]:
pipe1.named_steps['LASSO']

LogisticRegressionCV(Cs=20, cv=5, penalty='l1', solver='liblinear')

In [128]:
pipe1.named_steps['LASSO'].C_

array([0.08858668])

We can evaluate this like before:

In [129]:
## Score the Model on Training and Testing Set
result_scores['Logistic_SL1_C_auto'] = \
            (sklearn.metrics.accuracy_score(y_train,pipe1.predict(X_train)),
             sklearn.metrics.accuracy_score(y_test,pipe1.predict(X_test)))
get_results(result_scores)


Model                  Train    Test
-------------------------------------------
Logistic               0.7333   0.718 
Null                   0.6467   0.608 
Logistic_L1_C_1        0.732    0.716 
Logistic_L1_C_01       0.726    0.706 
Logistic_L1_C_10       0.7347   0.718 
Logistic_L1_C_auto     0.7233   0.708 
Logistic_SL1_C_auto    0.7307   0.714 


## Adding model complexity with interactions and polynomials

* We can add polynomials as well as interactions using PolynomialFeatures. 
* We must give the degree argument. Polynomials up to that degree will be considered, and interactions between d-1 terms.
* This is a lot of parameters added into the model! That is why we are using LASSO to shrink some and avoid overfitting…

In [130]:
#### This takes awhile to run! 
## try multiple polynomials with a LASSO regulaizer
## use pipeline for pre-processing
from sklearn.preprocessing import PolynomialFeatures

## set our transformation
scaler = preprocessing.StandardScaler()

## polynomials
poly_feat = PolynomialFeatures(degree=2,include_bias=False)

## set our model
clf = linear_model.LogisticRegressionCV(cv=3,
                                        Cs=5,
                                        penalty='l1',
                                        solver='liblinear') 
## put together in pipeline
pipe2 = Pipeline([("scale", scaler),
                  ("poly",poly_feat),
                  ("LASSO", clf)])
# pipe2.fit(X_train,y_train)

In [131]:
## Score the Model on Training and Testing Set
# result_scores['LogisticL1_SP_C_auto'] = \
#             (sklearn.metrics.accuracy_score(y_train,pipe1.predict(X_train)),
#              sklearn.metrics.accuracy_score(y_test,pipe1.predict(X_test)))
# get_results(result_scores)

## Random forest

- Another popular classification algorithm is the random forest. 
- In order to understand a random forest, you should first think about a decision tree. 
- We can consider modeling data simply by making cutpoints on our predictors, then splitting the decision of the outcome. 
- Visually, that might look like this in the context of our data:
- If age < 40, deny credit card
- If age > 40, then:
    + If income > 5, accept credit card
    + If income < 5, accept credit card
    
- Decision trees have a very nice appeal in that they are easy to understand and visualize. You can simply make a score card, and a human could easily make a decision on whether the outcome is yes or no. 
- But they are not very accurate or flexible individually! How can we have a good model from a single decision tree? It seems unlikely. 
- But what if we created many different decision trees, based on different subsets of the data? 
- We could take random samples of the data, then get an optimal decision tree using a subset of the predictors for each sample. 
- Each individual tree isn't that great, but perhaps the population of all those trees (the ensemble) would be good?
- That is the intuition behind a random forest!

The parameters of interest for tuning are `n_estimators` and `max_features`.   
  
`n_estimators` is the number of trees in the forest. Generally, the larger the number of tree, the better the prediction and more stable the algorithm. However, the algorihm takes longer the more trees we have. 
  
`max_features` is the maximum number of features (variables) to consider for each tree split. Not every tree will use every parameter. It is often suggested to use the square root of the number of features for classification here. 

In [132]:
#### Fit Random Forest
## Random Forests
from sklearn import ensemble
clf = ensemble.RandomForestClassifier(n_estimators=100, 
                                      max_features=10,
                                      random_state=42)
clf.fit(X_train,y_train)
## get confusion matrix
confusion_matrix(y_train,clf.predict(X_train))

RandomForestClassifier(max_features=10, random_state=42)

array([[970,   0],
       [  1, 529]])

In [133]:
## Score the Model on Training and Testing Set
result_scores['RandomForest_noCV'] = \
            (sklearn.metrics.accuracy_score(y_train,clf.predict(X_train)),
             sklearn.metrics.accuracy_score(y_test,clf.predict(X_test)))
get_results(result_scores)


Model                  Train    Test
-------------------------------------------
Logistic               0.7333   0.718 
Null                   0.6467   0.608 
Logistic_L1_C_1        0.732    0.716 
Logistic_L1_C_01       0.726    0.706 
Logistic_L1_C_10       0.7347   0.718 
Logistic_L1_C_auto     0.7233   0.708 
Logistic_SL1_C_auto    0.7307   0.714 
RandomForest_noCV      0.9993   0.684 


## Grid Search for Manual Cross Validation

There is no RandomForestCV function....what to do? 


We can specify a grid search across a range of hyperparameters. 

In [134]:
from sklearn.model_selection import GridSearchCV
## specify grid
parameters = {'n_estimators':(50,100,200,300),
              'max_features':(5,10,15,20)}
## specify model without hyperparameters
rf_model = ensemble.RandomForestClassifier(random_state=32)
## specify search with model
clf = GridSearchCV(rf_model,
                   parameters,
                   cv=5,
                   return_train_score=True)
clf.fit(X_train,y_train)

GridSearchCV(cv=5, estimator=RandomForestClassifier(random_state=32),
             param_grid={'max_features': (5, 10, 15, 20),
                         'n_estimators': (50, 100, 200, 300)},
             return_train_score=True)

In [135]:
## explore best hyperparameters
clf.best_params_

{'max_features': 10, 'n_estimators': 100}

In [136]:
## add model score
## Score the Model on Training and Testing Set
result_scores['RandomForest_CV'] = \
            (sklearn.metrics.accuracy_score(y_train,clf.predict(X_train)),
             sklearn.metrics.accuracy_score(y_test,clf.predict(X_test)))
get_results(result_scores)


Model                  Train    Test
-------------------------------------------
Logistic               0.7333   0.718 
Null                   0.6467   0.608 
Logistic_L1_C_1        0.732    0.716 
Logistic_L1_C_01       0.726    0.706 
Logistic_L1_C_10       0.7347   0.718 
Logistic_L1_C_auto     0.7233   0.708 
Logistic_SL1_C_auto    0.7307   0.714 
RandomForest_noCV      0.9993   0.684 
RandomForest_CV        0.9987   0.694 


In [137]:
clf.cv_results_

{'mean_fit_time': array([0.0439249 , 0.0877142 , 0.17621288, 0.26754036, 0.05440216,
        0.10772147, 0.216013  , 0.32144613, 0.06405406, 0.12943683,
        0.25546913, 0.3842083 , 0.0738009 , 0.14713526, 0.29611235,
        0.44330106]),
 'std_fit_time': array([0.00050443, 0.00104027, 0.0014419 , 0.00309228, 0.00080623,
        0.00110468, 0.00072777, 0.00211135, 0.00063561, 0.00178037,
        0.00233296, 0.00353177, 0.00098021, 0.00192756, 0.00245109,
        0.00484546]),
 'mean_score_time': array([0.00215535, 0.00407252, 0.00783715, 0.01170368, 0.00212193,
        0.00394745, 0.00771122, 0.01128774, 0.00205593, 0.00391002,
        0.00751553, 0.01117649, 0.00202613, 0.00383725, 0.00743961,
        0.01105051]),
 'std_score_time': array([9.50657603e-05, 1.63374362e-04, 1.00258752e-04, 1.55177010e-04,
        5.45273246e-05, 6.90015497e-05, 7.90805927e-05, 1.17265921e-04,
        1.08146037e-04, 9.40341011e-05, 1.47526250e-04, 1.86880300e-04,
        3.86242219e-05, 1.02933074e-

Let's add one more adjustment to the RandomForest. Since we are still overfitting, let's try optimizing the depth of the trees...

In [138]:
from sklearn.model_selection import GridSearchCV
## specify grid
parameters2 = {'max_depth':(2,5,7,10,20)}
## specify model without hyperparameters
rf_model = ensemble.RandomForestClassifier(max_features=20,
                                           n_estimators=100,
                                           random_state=32)
## specify search with model
clf = GridSearchCV(rf_model,
                   parameters2,
                   cv=5,
                   return_train_score=True)
clf.fit(X_train,y_train)

GridSearchCV(cv=5,
             estimator=RandomForestClassifier(max_features=20, random_state=32),
             param_grid={'max_depth': (2, 5, 7, 10, 20)},
             return_train_score=True)

In [139]:
## explore best hyperparameters
clf.best_params_

{'max_depth': 2}

In [140]:
## add model score
## Score the Model on Training and Testing Set
result_scores['RandomForest_CV2'] = \
            (sklearn.metrics.accuracy_score(y_train,clf.predict(X_train)),
             sklearn.metrics.accuracy_score(y_test,clf.predict(X_test)))
get_results(result_scores)


Model                  Train    Test
-------------------------------------------
Logistic               0.7333   0.718 
Null                   0.6467   0.608 
Logistic_L1_C_1        0.732    0.716 
Logistic_L1_C_01       0.726    0.706 
Logistic_L1_C_10       0.7347   0.718 
Logistic_L1_C_auto     0.7233   0.708 
Logistic_SL1_C_auto    0.7307   0.714 
RandomForest_noCV      0.9993   0.684 
RandomForest_CV        0.9987   0.694 
RandomForest_CV2       0.7267   0.702 


##### Q1 Among the different classification models included in the Python notebook, which model had the best overall performance? Support your response by referencing appropriate evidence.

Ans:  </br>

According to the results in the earlier code, the RandomForest_CV2 model performed the best overall on the holdout (test) set, with a test accuracy of 0.720. Although certain models, such as RandomForest_noCV and RandomForest_CV, showed almost perfect training accuracy (0.999), their test accuracy decreased dramatically (0.688 and 0.694, respectively), suggesting overfitting.

RandomForest_CV2, on the other hand, achieves a solid balance between training and test performance, with a training accuracy of 0.761 and the greatest test accuracy of any model. This shows that it generalizes more well than the other models.

As a result, according to test set performance, RandomForest_CV2 is the best performing model.




##### Q2 Next, fit a series of logistic regression models, without regularization. Each model should use the same set of predictors (all of the relevant predictors in the dataset) and should use the entire dataset, rather than a fraction of it. Use a randomly chosen 80% proportion of observations for training and the remaining for checking the generalizable performance (i.e., performance on the holdout subset). Be sure to ensure that the training and holdout subsets are identical across all models. Each model should choose a different solver.


##### Q3.Compare the results of the models in terms of their accuracy (use this as the performance metric to assess generalizability error on the holdout subset) and the time taken (use appropriate timing function). Summarize your results via a table with the following structure:

Solver used  | Training subset accuracy   | Holdout subset accuracy  | Time taken

In [141]:
# import necessary libraries
import time

# warning ignore
import warnings
warnings.filterwarnings('ignore')

# get the entire dataset
df_patient_whole = df_patient.copy()

## Create formula for all variables in model
vars_remove = ['PatientID','First_Appointment_Date','DateOfBirth',
               'Last_Appointment_Date','DateOfDeath','mortality']
vars_left_q2 = set(df_patient_whole.columns) - set(vars_remove)
formula_q2 = "mortality ~ " + " + ".join(vars_left_q2)
formula_q2
  
## use Patsy to create model matrices
Y,X = dmatrices(formula_q2,
                df_patient_whole)


'mortality ~ Gender + HIV + LiverMild + Congestive_heart_failure + Myocardial_infarction + Depression + Diabetes_without_complications + Peptic_ulcer_disease + Peripheral_vascular_disease + Renal + Pulmonary + Cancer + Diabetes_with_complications + Alcohol + Drugs + Rheumatic + LiverSevere + Stroke + Obesity + Dementia + Paralysis + Metastatic_solid_tumour + Age_years + Hypertension + Race'

In [142]:
## Split Data into training and sample

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = \
    train_test_split(X,
                     np.ravel(Y), # prevents dimensionality error later 
                     test_size=0.2, # 80% for training, 20% for testing
                     random_state=42)

In [143]:

# fit a series of logistic regression models, without regularization
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# define a list of solvers to try
solvers = ['newton-cg', 'lbfgs', 'liblinear', 'sag', 'saga']

# results list
results = []

# for each solver, fit a model and record the training and test accuracy
for solver in solvers:
    # to start the timer
    start_time = time.time()

    model = LogisticRegression(solver=solver)
    model.fit(X_train, y_train)

    # to stop the timer
    end_time = time.time()

    # to calculate the training time
    training_time = end_time - start_time

    # to calculate the training and test accuracy
    train_accuracy = model.score(X_train, y_train)
    test_accuracy = model.score(X_test, y_test)

    # to save the results
    results.append({
        'Solver used': solver,
        'Training subset accuracy': train_accuracy,
        'Holdout subset accuracy': test_accuracy,
        'Time taken': training_time
    })

# to create a dataframe from the results
results_df = pd.DataFrame(results)

# print out the results
print(results_df)

LogisticRegression(solver='newton-cg')

LogisticRegression()

LogisticRegression(solver='liblinear')

LogisticRegression(solver='sag')

LogisticRegression(solver='saga')

  Solver used  Training subset accuracy  Holdout subset accuracy  Time taken
0   newton-cg                  0.748188                  0.73625    0.032543
1       lbfgs                  0.748750                  0.73725    0.059317
2   liblinear                  0.747938                  0.73625    0.020731
3         sag                  0.748125                  0.73550    0.223584
4        saga                  0.748437                  0.73475    0.239353


##### Q4. Based on the results, which solver yielded the best results? Explain the basis for ranking the models - did you use training subset accuracy? Holdout subset accuracy? Time of execution? All three? Some combination of the three? </br>

Ans: </br>

The solution that demonstrated the most superior overall performance is lbfgs, as it got the maximum accuracy of 0.73725 on the holdout subset. In evaluating the models, the main standards was holdout accuracy, since it indicates the model's ability to generalize to novel data—an key facet of real-world predictive efficacy. Although training accuracy and execution time were evaluated, these were subordinate considerations. For instance, although liblinear had the briefest execution time, its holdout accuracy was marginally worse. Also, saga and sag had the greatest execution time and demonstrated worse performance on the holdout set. Therefore, by balancing performance and efficiency, lbfgs comes as the most efficient solution for this problem.